admin/cleanup.py

```diff
@@ -0,0 +1,379 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+"""
+Utilities for cloud resource cleanup.
+"""
+from datetime import datetime, timedelta
+import json
+import sys
+from uuid import UUID
+import yaml
+from yaml.parser import ParserError
+
+from characteristic import attributes
+
+from dateutil.parser import parse as parse_date
+from dateutil.tz import tzutc
+
+from libcloud.compute.providers import get_driver, Provider
+
+from twisted.internet.defer import succeed
+from twisted.python.filepath import FilePath
+from twisted.python.log import err
+from twisted.python.usage import Options, UsageError
+
+# Marker value defined by ``flocker.testtools.cluster_utils.MARKER``.  This
+# should never change and should always identify test-created clusters.
+from flocker.testtools.cluster_utils import MARKER
+
+
+def get_creation_time(node):
+    """
+    Get the creation time of a libcloud node.
+
+    Rackspace and EC2 store the information in different metadeta.
+
+    :return: The creation time, if available.
+    :rtype: datetime or None
+    """
+    date_string = node.extra.get("created", node.extra.get("launch_time"))
+    if date_string is None:
+        return None
+    else:
+        return parse_date(date_string)
+
+
+def get_rackspace_driver(rackspace):
+    """
+    Get a libcloud Rackspace driver given some credentials and other
+    configuration.
+    """
+    rackspace = get_driver(Provider.RACKSPACE)(
+        rackspace['username'], rackspace['key'],
+        region=rackspace['region'],
+    )
+    return rackspace
+
+
+def get_ec2_driver(aws):
+    """
+    Get a libcloud EC2 driver given some credentials and other configuration.
+    """
+    ec2 = get_driver(Provider.EC2)(
+        aws['access_key'], aws['secret_access_token'],
+        region=aws['region'],
+    )
+    return ec2
+
+
+def _get_volume_creation_time(volume):
+    """
+    Extract the creation time from an AWS or Rackspace volume.
+
+    libcloud doesn't represent volume creation time uniformly across
+    drivers.  Thus this method only works on drivers specifically accounted
+    for.
+    """
+    try:
+        # AWS
+        return volume.extra['create_time']
+    except KeyError:
+        # Rackspace.  Timestamps have no timezone indicated.  Manual
+        # experimentation indicates timestamps are in UTC (which of course
+        # is the only reasonable thing).
+        return parse_date(
+            volume.extra['created_at']
+        ).replace(tzinfo=tzutc())
+
+
+def _get_volume_region(volume):
+    """
+    Get the name of the region the volume is in.
```

Comment:
:question: If you choose not to put this behind an interface, could you change this comment to something more like the one above?

Specifically, note that this is "best effort" or "Only works for AWS and Rackspace" and maybe also grumble about libcloud not actually presenting a uniform interface to volume data.

Looks like this is just used for logging on `--dry_run` so best effort is probably fun, but perhaps even call that out with a `N.B.`